README
Next Generation Init System ( initng )
======================================

Contents
--------
1    Introduction
  1.1   Developers
2    Initng: An Overview
  2.1   The Next Generation Control (ngc)
3    How to get initng
  3.1   How to Install
  3.2   Caveats/Known Issues
4    Terms of Licensing
5    Development



1 Introduction
--------------

Initng is a full replacement of the old and in many ways deprecated sysvinit
tool. It is designed with speed in mind, doing as much as possible
asynchronously. In other words: It will boot your unix-system much faster, and
give you more control and statistics over your system.

The basic premise is that startup commands can be launched as soon as their
dependencies are met. This limits the effect of bottlenecks like I/O operations;
while one program is performing I/O, another can be utilizing the processor.
Initng tracks the individual service dependencies in its configuration files. 

It is designed to use a minimum of system resources and to boot your system
quickly and safely.


Website:        http://www.initng.org
Mailing list:   http://www.initng.org/wiki/Contact_Lists


This package currently includes:

- initng       sysvinit replacement; handles system & service control
- ngc          an application for controlling initng
- plugins      initng uses plugins to provide some functions


1.1 Developers
--------------

If you find this little piece of software useful, please email me
<jimmy.wennlund@gmail.com> with some feedback.


* Jimmy Wennlund <jimmy.wennlund@gmail.com>

  I am the original author of the initng project, which was created in my spare
time and mostly for fun. I live in Sweden, outside Uppsala. 



2 Initng: An Overview
---------------------

On boot, initng will be started as the first process (pid = 1) by the kernel.
At first, initng will parse configuration files located in /etc/initng for
critical information such as runlevel and service data. After that, all services
required by the default runlevel will be started as soon as their dependencies
are met, allowing services to run in parallel.  This asynchronous execution can
dramatically improve boot time by better utilizing the system resources
(especially in the case of multiprocessor systems).

Service monitoring is also available, automatically respawing daemons that die
without being explicitly shutdown, a critical feature for server systems that
require constant uptime.

The system gives complete control over how your system boots in the traditional
services and runlevels model.  Startup status is presented in a percentile
fashion, as well as the launch time for each individual service.  This allows
easier tracking of what services add the most time to startup.



2.1 The Next Generation Control (ngc)
-------------------------------------

Using the included ngc control program to communicate with initng in a two-way
communication, checking the status of all services, browse history, bring
services up and down, reboot or halt, and change runlevels.

Executing 'ngc -h' will display the command's help and description.  'ngc -H'
displays extended help.



3 How to get initng
-------------------

For now, a test or relay first beta is located at http://jw.dyndns.org/initng/.
You are free to test AT YOUR OWN RISK! REMEMBER, this is a beta release
and many new features that are on the TODO list are not yet fixed.


3.1 How to Install
-----------------

1. Download the latest version from www.initng.org

2. Unzip the tarball, enter the directory, and install as super-user

    # tar -zxvf initng-<version>.tar.gz
    # cd initng-<version>
	# cmake
    # make install

3. Instruct your bootloader to use the initng instead of the traditional 'init'

    WARNING: initng is beta software, and must be used with caution. However,
    it can be installed side-by-side with init. It is recommended that you
    duplicate the relevant blocks of your bootloader configuration to add the
    above changes. That is, copy your existing entry (eg, the one labeled
    "linux") to a new section (say, "linux-initng") before you add the "init="
    line. This way you can always boot with the original (default) init.

* LILO
    Edit lilo.conf and add the line append="init=/sbin/initng" after the
    image specifier. 

    # Sample LILO entry
    image = /boot/vmlinuz-2.6.9
      root = /dev/hda6
      label = linux-initng
      append = "init=/sbin/initng"
      read-only


* GRUB
    Edit /etc/grub/grub.conf and append init=/sbin/initng to the image specifier

    # Sample GRUB entry
    title  linux-initng
    kernel (hd0,1)/vmlinuz root=/dev/hda6 init=/sbin/initng


[Note: 'make uninstall' is available]



3.2 Caveats/Known Issues
------------------------

initng is beta software - expect incomplete features and possible instability.
Use this piece of software with care; it is not yet recommended for use on
production platforms.

This program controls many low level systems, including partition related
services.  Back up your data before you use it.


4 Terms of Licensing
--------------------

The full package is licensed under the GNU Public License (GPL) version 2,
released in June of 1991.  The full text of the GPL is located in this package
and can be retireved from http://www.gnu.org/copyleft/gpl.html in a variety of
formats.



5 Development
-------------

Contact the mailing list: http://jw.dyndns.org/mailman/listinfo/initng

Jimmy Wennlund
	<jimmy.wennlund@gmail.com>
	+46 737 623932



NEWS
Version History:
---------------
2007/03/25
initng-0.6.10 Adesso e Fortuna
	* Disabled the find plugin, it's problematic.
	* Added RUNLEVEL_PREFIX option, defaults to "runlevel/".
	* Set close-on-exec flag on the opened files to avoid explicitely
	  closing them.
	* Finished the service_file plugin, it's usable now.

2007/02/03
initng-0.6.9 Shappire Sparks
	* Implemented an internal event system, that replaces plugin-hooks.
	* Added event plugin, the skeleton for ifiles event support.
	* Added FORCE_POSIX_IFILES option, to force POSIX compliance on ifiles.
	* Changed the beviahour of initng when pid != 1, now fake-mode is activated via a parameter (--fake)
	* Updated the documentation.
	* A lot of bugfixing.

2006/09/19
initng-0.6.8
	* Optimized the depend algorithm.
	* Automatically zap failed services if the source changes.
	* Added service_file plugin (experimental).
	* Some bug fixes.

2006/04/18
initng-0.6.7 Never been so good (Part 4)
	* Code listed by accident under the LGPL is now GPL again
	* Added netdev plugin, network interfaces can now be specified using need = device/eth0;
	* Added fmon plugin, if initng sources are modified, initng will notice.
	* Plugin paths are saved in initng core for monitoring.
	* killalli5 has been improved with an ignore list, tries not to kill initng services.
	* initng-usplash plugin now provides proper usplash support with ubuntu.
	* Plugins are now loaded in alpabetical order, might prevent bugs when the order is random, also easier to test.
	* New version check when loading plugins, hopefully no more problems where stale plugins in /lib/initng* prevented your computer from booting.
	* Thanks to fmon plugin, hopefully problems mounting / read only due to a kernel bug are fixed, read wiki FAQ for more info about this bug.
	* nge plugin now gives some more info to the listeners.
	* nge protocol version bumped.
	* add cpout_nocolors to initline if you dont like ansi colors on screen.
	* Lots of bugfixing as usual.

2006/05/12
initng-0.6.6 Never been so good (Part 3)
	* In service type, the services can be set as hidden, for more internal services used as deps only.
	* cpout plugin will now not print services marked as hidden.
	* ngc -s wont print hidden services.
	* ngc -S is moved to ngc -t
	* ngc -S will work like ngc -s, but also show hidden services.
	* provide plugin got some fixes.
	* selinux is hopefully working now, please test and report.
	* Some bugfixing all over.
	* set cpout_console:/dev/tty10 in grub to redirect the nice cpout console somewhere else.

initng-0.6.5 Never been so good (Part 2)
	* A bug in iparser made initng segfault if trying to start a service like test/*/* with ngc -u test/abc/def.
	* Add a state REQ_NOT_FOUND that will be put on services that require = not_found; instead of refusing to start them.
	* Selinux code moved to a plugin, is so far untested.

2006/05/06
initng-0.6.4 Never been so good
	* GNU autotools support has been removed completely - cmake is better in many ways.
	* Install the libs to /lib64 if needed.
	* The nge socket now reports % complete on boot, is possible to build splash daemons listening on ngebus during boot.
	* Work on making escape chars work in i-files, with i-parser. \; wont break the row anymore.
	* Many string functions were rewritten to decrease cpu usage.
	* Added the possibility to use test/*/* services.
	* Fixed a cirular bug in i_parser used by the find plugin that could have locked up initng.
	* Added a new command to be able to manually launch scripts in i-files for other functions than starting and stopping a service.
	* Fixed a buffer handling bug.
	* Ngc4 protocol updates.
	* Add state descriptions on all states printed by ngc -S (--states).
	* Fixed random ngc command failure.
	* Lots of bugfixing and memleaks fixed as usual.
	* Add a provide plugin for service virtuals (not enabled by default to copile, this plugin is pretty untested).
	* Add a NO_CACHE option which can be added to service_types which are not based on a file. This one is experimental, and only used by the provide plugin today which is not compiled by default.
	* Ngc4 header updated.

2006/03/26 12:00 CET
initng-0.6.3
	* Added Selinux support.
	* Rewrote fix_variables, the new version is faster, but may still contain bugs.
	* Fix_varibles rewrite, env_parser = $CONFFILE; now works.
	* Fixed bug where if A=HELLO, B=$A, C=$B, C was not assigned the correct value (#553).
	* Fixed bug where env_file was not always parsed correctly.
	* Add ngc -S (--states) which lists the current states services can have, and soon maybe description of the states.

2006/03/24 15:00 CET
initng-0.6.2
	* Fixed a bug in the ngc4 plugin; file descriptors were not closed and could cause remounting / read-only on shutdown to fail
	* Non-essential console output has been reduced
	* Nge plugin partially redesigned
	* Core functions renamed to better describe their purpose

2006/03/21 12:00 CET
initng-0.6.1 Progressing fast
	* Completed support for cmake - you will need cmake to install Initng now!!!  Autotools is no longer supported and if you use it, you do so at your own risk
	* New website adress http://www.initng.org/
	* Only one runlevel (note system is no longer counted as a runlevel - you must rename it to system.virtual) can be up at once, trying to start a second runlevel will stop the first, along with any services not listed in the new runlevel
	* add itype binary, that will soon be a tool for offline usage of .i files
	* Make ngc -u /etc/full/path/ifile.i work
	* Rewrote process buffer handling
	* Filecache updates: Initng won't read an i file over and over again for each entry in it
	* Completely rewritten ngc4 plugin: nicer output, written as an clientpart so writing gui-tools using ngc is now possible.
	* Added comments to the code
	* Documentation/templates added for all plugins
	* Cleanups in the code and output
	* Fixed segfault from DBus plugin
	* Initng data storage backend rewritten.

src/main.c
/*
 * Initng, a next generation sysvinit replacement.
 * Copyright (C) 2006 Jimmy Wennlund <jimmy.wennlund@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include "initng.h"

#include <unistd.h>
#include <time.h>							/* time() */
#include <fcntl.h>							/* fcntl() */
#include <linux/kd.h>						/* KDSIGACCEPT */
#include <stdlib.h>							/* free() exit() */
#include <termios.h>
#include <stdio.h>
#include <dirent.h>
#include <fnmatch.h>
#include <errno.h>
#include <ctype.h>

#include <sys/stat.h>
#include <sys/klog.h>
#include <sys/reboot.h>						/* reboot() RB_DISABLE_CAD */
#include <sys/ioctl.h>						/* ioctl() */
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/un.h>							/* memmove() strcmp() */
#include <sys/wait.h>						/* waitpid() sa */
#include <sys/mount.h>
#include <initng-paths.h>

#include "initng_global.h"
#include "initng_signal.h"
#include "initng_handler.h"
#include "initng_execute.h"					/* new_environ() */
#include "initng_active_db.h"
#include "initng_load_module.h"
#include "initng_plugin_callers.h"
#include "initng_toolbox.h"
#include "initng_interrupt.h"
#include "initng_main.h"
#include "initng_service_data_types.h"
#include "initng_fd.h"
#include "initng_process_db.h"
#include "initng_global.h"
#include "initng_static_event_types.h"

#ifdef SELINUX
#include <stdlib.h>
#include <string.h>
#include <selinux/selinux.h>
#include <selinux/context.h>
#include <sepol/sepol.h>
#endif
#define TIMEOUT 60000


/*older code no longer needed on FC5 and FCX (X>=5) */
#ifdef OLDSELINUX
/* Mount point for selinuxfs. */
#define SELINUXMNT "/selinux/"
int enforcing = -1;

static int load_policy(int *enforce)
{
	int fd = -1, ret = -1;
	size_t data_size;
	int rc = 0, orig_enforce;
	struct stat sb;
	void *map, *data;
	char policy_file[PATH_MAX];
	int policy_version = 0;
	FILE *cfg;
	char buf[4096];
	int seconfig = -2;
	char *nonconst;				//Ugly hack!

	selinux_getenforcemode(&seconfig);

	mount("none", "/proc", "proc", 0, 0);
	cfg = fopen("/proc/cmdline", "r");
	if (cfg)
	{
		char *tmp;

		if (fgets(buf, 4096, cfg) && (tmp = strstr(buf, "enforcing=")))
		{
			if (tmp == buf || isspace(*(tmp - 1)))
			{
				enforcing = atoi(tmp + 10);
			}
		}
		fclose(cfg);
	}
#define MNT_DETACH 2
	umount2("/proc", MNT_DETACH);

	if (enforcing >= 0)
		*enforce = enforcing;
	else if (seconfig == 1)
		*enforce = 1;

	if (mount("none", SELINUXMNT, "selinuxfs", 0, 0) < 0)
	{
		if (errno == ENODEV)
		{
			fprintf(stderr, "SELinux not supported by kernel: %s\n",
					strerror(errno));
			*enforce = 0;
		}
		else
		{
			fprintf(stderr, "Failed to mount %s: %s\n", SELINUXMNT,
					strerror(errno));
		}
		return ret;
	}

	nonconst = malloc(sizeof(SELINUXMNT));
	strcpy(nonconst, SELINUXMNT);
	set_selinuxmnt(nonconst);				/* set manually since we mounted it */
	free(nonconst);

	policy_version = security_policyvers();
	if (policy_version < 0)
	{
		fprintf(stderr, "Can't get policy version: %s\n", strerror(errno));
		goto UMOUNT;
	}

	orig_enforce = rc = security_getenforce();
	if (rc < 0)
	{
		fprintf(stderr, "Can't get SELinux enforcement flag: %s\n",
				strerror(errno));
		goto UMOUNT;
	}
	if (enforcing >= 0)
	{
		*enforce = enforcing;
	}
	else if (seconfig == -1)
	{
		*enforce = 0;
		rc = security_disable();
		if (rc == 0)
			umount(SELINUXMNT);
		if (rc < 0)
		{
			rc = security_setenforce(0);
			if (rc < 0)
			{
				fprintf(stderr, "Can't disable SELinux: %s\n",
						strerror(errno));
				goto UMOUNT;
			}
		}
		ret = 0;
		goto UMOUNT;
	}
	else if (seconfig >= 0)
	{
		*enforce = seconfig;
		if (orig_enforce != *enforce)
		{
			rc = security_setenforce(seconfig);
			if (rc < 0)
			{
				fprintf(stderr, "Can't set SELinux enforcement flag: %s\n",
						strerror(errno));
				goto UMOUNT;
			}
		}
	}

	snprintf(policy_file, sizeof(policy_file), "%s.%d",
			 selinux_binary_policy_path(), policy_version);
	fd = open(policy_file, O_RDONLY);
	if (fd < 0)
	{
		/* Check previous version to see if old policy is available
		 */
		snprintf(policy_file, sizeof(policy_file), "%s.%d",
				 selinux_binary_policy_path(), policy_version - 1);
		fd = open(policy_file, O_RDONLY);
		if (fd < 0)
		{
			fprintf(stderr, "Can't open '%s.%d':  %s\n",
					selinux_binary_policy_path(), policy_version,
					strerror(errno));
			goto UMOUNT;
		}
	}

	if (fstat(fd, &sb) < 0)
	{
		fprintf(stderr, "Can't stat '%s':  %s\n",
				policy_file, strerror(errno));
		goto UMOUNT;
	}

	map = mmap(NULL, sb.st_size, PROT_READ | PROT_WRITE, MAP_PRIVATE, fd, 0);
	if (map == MAP_FAILED)
	{
		fprintf(stderr, "Can't map '%s':  %s\n",
				policy_file, strerror(errno));
		goto UMOUNT;
	}


	/* Set SELinux users based on a local.users configuration file. */
	ret = sepol_genusers(map, sb.st_size, selinux_users_path(), &data,
						 &data_size);
	if (ret < 0)
	{
		fprintf(stderr,
				"Warning!  Error while reading user configuration from %s/{local.users,system.users}:  %s\n",
				selinux_users_path(), strerror(errno));
		data = map;
		data_size = sb.st_size;
	}

	/* Set booleans based on a booleans configuration file. */
	nonconst = malloc(sizeof(selinux_booleans_path()));
	strcpy(nonconst, selinux_booleans_path());
	ret = sepol_genbools(data, data_size, nonconst);
	free(nonconst);
	if (ret < 0)
	{
		if (errno == ENOENT || errno == EINVAL)
		{
			/* No booleans file or stale booleans in the file; non-fatal. */
			fprintf(stderr, "Warning!  Error while setting booleans:  %s\n",
					strerror(errno));
		}
		else
		{
			fprintf(stderr, "Error while setting booleans:  %s\n",
					strerror(errno));
			goto UMOUNT;
		}
	}
	fprintf(stderr, "Loading security policy\n");
	ret = security_load_policy(data, data_size);
	if (ret < 0)
	{
		fprintf(stderr, "security_load_policy failed\n");
	}

  UMOUNT:
	/*umount(SELINUXMNT); */
	if (fd >= 0)
		close(fd);

	return (ret);
}
#endif


static void setup_console(void)
{
	int fd;						/* /dev/console */
	struct termios tty;

	D_("MAIN_SET_I_AM_INIT_STUFF\n");

	/* change dir to / */
	while (chdir("/") < 0)
	{
		F_("Cant chdir to \"/\"\n");
		initng_main_su_login();
	}

	/* set console loglevel */
	klogctl(8, NULL, 1);


	/* enable generation of core files */
	{
		struct rlimit c = { 1000000, 1000000 };
		setrlimit(RLIMIT_CORE, &c);
	}

	reboot(RB_DISABLE_CAD);					/* Disable Ctrl + Alt + Delete */

	/* open the console */
	if (g.dev_console)
	{
		fd = open(g.dev_console, O_RDWR | O_NOCTTY);
	}
	else
	{
		fd = open(INITNG_CONSOLE, O_RDWR | O_NOCTTY);
	}

	/* Try to open the console, but don't control it */
	if (fd > 0)
	{
		D_("Opened " INITNG_CONSOLE ". Setting terminal options.\n");
		ioctl(fd, KDSIGACCEPT, SIGWINCH);	/* Accept signals from 'kbd' */
		close(fd);							/* Like Ctrl + Alt + Delete signal? */
	}
	else
	{
		D_("Failed to open " INITNG_CONSOLE ". Setting options anyway.\n");
		ioctl(0, KDSIGACCEPT, SIGWINCH);	/* Accept signals from 'kbd' */
	}

	/* TODO: this block may be incorrect or incomplete */
	/* Q: What does this block really do? */

	/*
	 * TODO: /dev/console may still be open from before. Also, if it
	 * fails to open, why try to finish the block?
	 */
	if ((fd = open("/dev/console", O_RDWR | O_NOCTTY)) < 0)
	{
		F_("main(): can't open /dev/console.\n");
	}

	(void) tcgetattr(fd, &tty);

	tty.c_cflag &= CBAUD | CBAUDEX | CSIZE | CSTOPB | PARENB | PARODD;
	tty.c_cflag |= HUPCL | CLOCAL | CREAD;

	tty.c_cc[VINTR] = 3;					/* ctrl('c') */
	tty.c_cc[VQUIT] = 28;					/* ctrl('\\') */
	tty.c_cc[VERASE] = 127;
	tty.c_cc[VKILL] = 24;					/* ctrl('x') */
	tty.c_cc[VEOF] = 4;						/* ctrl('d') */
	tty.c_cc[VTIME] = 0;
	tty.c_cc[VMIN] = 1;
	tty.c_cc[VSTART] = 17;					/* ctrl('q') */
	tty.c_cc[VSTOP] = 19;					/* ctrl('s') */
	tty.c_cc[VSUSP] = 26;					/* ctrl('z') */

	/*
	 *      Set pre and post processing
	 */
	tty.c_iflag = IGNPAR | ICRNL | IXON | IXANY;
	tty.c_oflag = OPOST | ONLCR;
	tty.c_lflag = ISIG | ICANON | ECHO | ECHOCTL | ECHOPRT | ECHOKE;

	/*
	 *      Now set the terminal line.
	 *      We don't care about non-transmitted output data
	 *      and non-read input data.
	 */
	(void) tcsetattr(fd, TCSANOW, &tty);
	(void) tcflush(fd, TCIOFLUSH);
	(void) close(fd);

}


/*
 * %%%%%%%%%%%%%%%%%%%%   main ()   %%%%%%%%%%%%%%%%%%%%
 */
#ifdef BUSYBOX
int init_main(int argc, char *argv[], char *env[])
{
#else
int main(int argc, char *argv[], char *env[])
{
#endif
	struct timeval last;		/* save the time here */
	int retval;

#ifdef DEBUG
	int loop_counter = 0;		/* counts how many times the main_loop has run */
#endif
	S_;

	/* maby initng is launched only for getting the version */
	if (argv[1] && strcmp(argv[1], "--version") == 0)
	{
		fprintf(stderr, INITNG_VERSION "\n");
		usleep(100);
		fprintf(stdout, "api_ver=%i.\n", API_VERSION);
		fprintf(stdout, "created by " INITNG_CREATOR "\n\n");
		exit(0);
	}

	/* get the time */
	gettimeofday(&last, NULL);


	/* initialize global variables */
	initng_global_new(argc, argv, env);

	if (getuid() != 0)
	{
		W_("Initng is designed to run as root user, a lot of functionality will not work correctly.\n");
	}

	/* if this is real init */
	if (g.i_am == I_AM_INIT)
	{
		/*load selinux policy and rexec*/
#ifdef SELINUX
		FILE *tmp_f;
		if ((tmp_f = fopen("/selinux/enforce", "r")) != NULL)
		{
			fclose(tmp_f);
			goto BOOT;
		}

		int enforce = -1;
		char *nonconst;

		if (getenv("SELINUX_INIT") == NULL)
		{
			nonconst = malloc(sizeof("SELINUX_INIT=YES"));
			strcpy(nonconst, "SELINUX_INIT=YES");
			putenv(nonconst);
			free(nonconst);
#ifdef OLDSELINX
			if (load_policy(&enforce) == 0)
#else
			if (selinux_init_load_policy(&enforce) == 0)
#endif
			{
				execv(g.Argv[0], g.Argv);
			}
			else
			{
				if (enforce > 0)
				{
					/* SELinux in enforcing mode but load_policy failed */
					/* At this point, we probably can't open /dev/console, so log() won't work */
					fprintf(stderr,
						"Enforcing mode requested but no policy loaded. Halting now.\n");
					exit(1);
				}
			}
		}

		BOOT:
#endif

		/* when last service stopped, offer a sulogin */
		g.when_out = THEN_SULOGIN;
		if (!g.runlevel)
			initng_main_set_runlevel(RUNLEVEL_PREFIX "default");

		if (!g.hot_reload)
		{
			/* static function above, initialize the system */
			setup_console();
		}

	}
	else if (g.i_am == I_AM_FAKE_INIT)
	{
		/* when last service stopped, quit initng */
		W_("Initng is running in fake-mode, fake-default runlevel will be started instead.\n");
		g.when_out = THEN_QUIT;
		initng_main_set_runlevel(RUNLEVEL_PREFIX "fake-default");
	}

	D_("MAIN_LOAD_MODULES\n");
	/* Load modules, if fails - launch sulogin and then try again */
	if (!initng_load_module_load_all(INITNG_PLUGIN_DIR)
		&& !initng_load_module_load_all("/lib/initng")
		&& !initng_load_module_load_all("/lib32/initng")
		&& !initng_load_module_load_all("/lib64/initng"))
	{
		/* if we can't load modules, revert to single-user login */
		initng_main_su_login();
	}

	if (g.hot_reload)
	{
		/* Mark service as up */
		retval = initng_plugin_callers_reload_active_db();
		if (retval != TRUE)
		{
			if (retval == FALSE)
				F_("No plugin handled hot_reload!\n");
			else
				F_("Hot reload failed!\n");

			/* if the hot reload failed, we're probably in big trouble */
			initng_main_su_login();
		}

		/* Hopefully no-one will try a hot reload when the system isn't up... */
		initng_main_set_sys_state(STATE_UP);
	}

	/* set title of initng, can be watched in ps -ax */
	initng_toolbox_set_proc_title("initng [%s]", g.runlevel);

	/* Configure signal handlers */
	(void) initng_signal_enable();


	/* make sure this is not a hot reload */
	if (!g.hot_reload)
	{
		/* change system state */
		initng_main_set_sys_state(STATE_STARTING);

		/* first start all services, prompted at boot with +daemon/test */
		initng_main_start_extra_services();

		/* try load the default service, if it fails - launch sulogin and try again */
		if (!initng_handler_start_new_service_named(g.runlevel))
		{
			F_("Failed to load runlevel (%s)!\n", g.runlevel);
			initng_main_su_login();
		}
	}

	D_("MAIN_GOING_MAIN_LOOP\n");
	/* %%%%%%%%%%%%%%%   MAIN LOOP   %%%%%%%%%%%%%%% */
	for (;;)
	{
		D_("MAIN_LOOP: %i\n", loop_counter++);
		int interrupt = FALSE;

		/* Update current time, save this in global so we don't need to call time() that often. */
		gettimeofday(&g.now, NULL);

		/*
		 * If it has gone more then 1h since last
		 * main loop, or if clock has gone backwards,
		 * reset it.
		 */
		if ((g.now.tv_sec - last.tv_sec) >= 3600
			|| (g.now.tv_sec - last.tv_sec) < 0)
		{
			D_(" Clock skew, time have changed over one hour, in one mainloop, compensating %i seconds.\n", (g.now.tv_sec - last.tv_sec));
			initng_active_db_compensate_time(g.now.tv_sec - last.tv_sec);
			initng_plugin_callers_compensate_time(g.now.tv_sec - last.tv_sec);
			last.tv_sec += (g.now.tv_sec - last.tv_sec);
		}
		/* put last time, to current time last = g.now; */
		memcpy(&last, &g.now, sizeof(struct timeval));

		/* if a sleep is set */
		g.sleep_seconds = 0;

		/* run state alarm if timeout */
		/*
		 * Run all alarm state callers.
		 * If there is more later alarms, this will set new ones.
		 */
		if (g.next_alarm && g.next_alarm <= g.now.tv_sec)
			initng_handler_run_alarm();

		/* handle signals */
		{
			int i;

			for (i = 0; i < SIGNAL_STACK; i++)
			{
				if (signals_got[i] == -1)
					continue;

				initng_plugin_callers_signal(signals_got[i]);

				switch (signals_got[i])
				{
						/* dead children */
					case SIGCHLD:
						initng_signal_handle_sigchild();
						break;
					case SIGALRM:
						/*initng_plugin_callers_alarm(); */
						initng_handler_run_alarm();
						break;
					default:
						break;
				}
				/* reset the signal */
				signals_got[i] = -1;
			}

		}


		/* If there is modules to unload, handle this */
		if (g.modules_to_unload == TRUE)
		{
			g.modules_to_unload = FALSE;
			D_("There is modules to unload!\n");
			initng_unload_module_unload_marked();
		}


		/*
		 * After here i put things that should be done every main
		 * loop, g.interrupt or not ...
		 */

		/*
		 * Trigger MAIN_EVENT event.
		 * OBS! This part in mainloop is expensive.
		 */
		{
			s_event event;
			event.event_type = &EVENT_MAIN;
			initng_event_send(&event);
		}


		/*
		 * Check if there are any running processes left, otherwise quit, when_out try to do something when this happens.
		 *
		 * This check is also expensive.
		 */
		if (initng_main_ready_to_quit() == TRUE)
		{
			initng_main_set_sys_state(STATE_ASE);
			P_(" *** Last service has quit. ***\n");
			initng_main_when_out();
			continue;
		}

		/*
		 * Run this to clean the active_db out from down services.
		 */
		initng_active_db_clean_down();

		/* LAST: check for service state changes in a special function */
		interrupt = initng_interrupt();

		/* figure out how long we can sleep */
		{
			/* set it to default */
			int closest_timeout = TIMEOUT;

			/* if sleepseconds are set */
			if (g.sleep_seconds && g.sleep_seconds < closest_timeout)
				closest_timeout = g.sleep_seconds;

			/* Check how many seconds there are to the state alarm */
			if (g.next_alarm && g.next_alarm > g.now.tv_sec)
			{
				int time_to_next = g.next_alarm - g.now.tv_sec;

				D_("g.next_alarm is set!, will trigger in %i sec.\n",
				   time_to_next);
				if (time_to_next < closest_timeout)
					closest_timeout = time_to_next;
			}

			/* if we got some seconds */
			if (closest_timeout > 0 && interrupt == FALSE)
			{
				/* do a poll == the same as sleep but also watch fds */
				D_("Will sleep for %i seconds.\n", closest_timeout);
				initng_fd_plugin_poll(closest_timeout);
			}
		}
	}										/* End main loop */
}
